dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java

```diff
@@ -2100,32 +2170,125 @@ public void updateStepReferences(final String stepId, final String replacementSt
 	}
 
 	/**
-	 * Filter the list of actions to display according to the role passed
-	 * @param actions
-	 * @param role
-	 * @return
+	 * Filters the list of actions to display according to the role passed
+	 * @param actions the actions to filter by role
+	 * @param role the role selected
+	 * @return the list of actions once the filter operation has taken place
 	 * @throws DotDataException
 	 */
 	@CloseDBIfOpened
-	private List<WorkflowAction> filterActionsCollection(final List<WorkflowAction> actions,
-			final Role role) throws DotDataException {
+	private List<WorkflowAction> filterActionsCollection(final List<WorkflowAction> actions, final Role role) throws DotDataException {
 
 		final List<WorkflowAction> permissionables = new ArrayList<>(actions);
 		if (permissionables.isEmpty()) {
 			return permissionables;
 		}
 
+		//First try to determine if we're dealing with a role directly mapped to a user.
+		try {
+			final User user = userAPI.loadUserById(role.getRoleKey());
+			return filterActionsCollection(permissionables, user,true, null);
+		}catch (Exception nsu){
+			Logger.debug(this, () -> "Unable to determine role belongs to a user.");
+		}
+
 		final Iterator<WorkflowAction> workflowActionIterator = permissionables.iterator();
+
+		final List<Role> specialRolesHierarchy = getSpecialRolesHierarchy();
+		final int index = specialRolesHierarchy.indexOf(role);
+		//Determine if we're dealing with a special role
+		if(index >= 0){
+			//If so.. get a sublist with preserving the precedence. so we can apply a filter with the resulting piece of hierarchy.
+			final List<Role> rolesSubset = specialRolesHierarchy.subList(index, specialRolesHierarchy.size());
+			while (workflowActionIterator.hasNext()) {
+				final WorkflowAction wa = workflowActionIterator.next();
+				if (!hasAnyRolePermission(wa, rolesSubset)) {
+					workflowActionIterator.remove();
+				}
+			}
+			return permissionables;
+		}
+
+		//Perform filter for regular roles
+		final Set<Role> roles = collectChildRoles(role);
 		while (workflowActionIterator.hasNext()) {
 			final WorkflowAction wa = workflowActionIterator.next();
-			if (!permissionAPI.doesRoleHavePermission(wa, PermissionAPI.PERMISSION_USE, role)) {
+			if (!hasAnyRolePermission(wa, roles)) {
 				workflowActionIterator.remove();
 			}
 		}
 
 		return permissionables;
 	}
 
+
+	/**
+	 * Returns an immutable list with the Hierarchy of special roles according to the permission precedence.
+	 * This means that the position roles[i] has more precedence in that roles[i+1]
+	 * @return a list of special roles ordered by permission precedence.
+	 * @throws DotDataException
+	 */
+	private synchronized List<Role> getSpecialRolesHierarchy() throws DotDataException {
+
+		if (specialRolesHierarchy == null) {

```

Comment:
I would use the double check algorithm

if (x == null) {
   sync {
     if (x == null) {

I would also use a java.util.Set

Reply:
I'm Keeping the List as order matters and it only has 4 elements

Reply:
double check also ... done